File: DIP/Spoilers/MarsRoverKata/Behaviors/Tracker.cs
using MarsRoverKata.Commands;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MarsRoverKata.Behaviors
{
    public class Tracker : IBehavior
    {
        private IMovable Target;
        private int Speed;
        public Alien Parent { get; set; }

        public Tracker(IMovable thingToTrack)
        {
            Target = thingToTrack;
            Speed = 3;
        }

        public void ExecuteBehavior()
        {
            for (int i = 0; i < Speed; i++)
            {
                var delta = Target.Location - Parent.Location;

                var targetDirection = directionToTarget(delta);

                var success = Parent.Facing == targetDirection ? Parent.MoveForward() : Parent.TurnRight(); // new ForwardCommand(Parent).Execute() : new TurnRightCommand(Parent).Execute();
            }
        }

        private Direction directionToTarget(Point delta)
        { 
            var absoluteX = Math.Abs(delta.X);
            var absoluteY = Math.Abs(delta.Y);
            var greatestDelta = absoluteX > absoluteY ? delta.X : delta.Y;
            
            bool isNegative = greatestDelta < 0;
            
            return greatestDelta == delta.X ? determineEastOrWest(isNegative) : determineNorthOrSouth(isNegative);
        }

        private Direction determineEastOrWest(bool isNegative)
        {
            return isNegative ? Direction.West : Direction.East;
        }

        private Direction determineNorthOrSouth(bool isNegative)
        {
            return isNegative ? Direction.South : Direction.North;
        }
    }
}

File: DIP/Java_Final/MarsRoverKata/src/main/java/MarsRoverKata/Behaviors/Tracker.java
package MarsRoverKata.Behaviors;

import MarsRoverKata.Alien;
import MarsRoverKata.Direction;
import MarsRoverKata.IMovable;
import MarsRoverKata.Point;

public class Tracker implements IBehavior {
    private IMovable target;

    public IMovable getTarget() { return target; }

    private int Speed;
    private Alien parent;

    public Alien getParent() { return parent; }

    public void setParent(Alien parent) { this.parent = parent; }

    public Tracker(IMovable thingToTrack)
    {
        this.target = thingToTrack;
        Speed = 3;
    }

    public void executeBehavior()
    {
        for (int i = 0; i < Speed; i++)
        {
            Point delta = Point.subtract(getTarget().getLocation(), getParent().getLocation());

            Direction targetDirection = directionToTarget(delta);

            boolean success = getParent().getFacing() == targetDirection ? getParent().moveForward() : getParent().turnRight(); // new ForwardCommand(Parent).Execute() : new TurnRightCommand(Parent).Execute();
        }
    }

    private Direction directionToTarget(Point delta)
    {
        int absoluteX = Math.abs(delta.getX());
        int absoluteY = Math.abs(delta.getY());
        int greatestDelta = absoluteX > absoluteY ? delta.getX() : delta.getY();

        boolean isNegative = greatestDelta < 0;

        return greatestDelta == delta.getX() ? determineEastOrWest(isNegative) : determineNorthOrSouth(isNegative);
    }

    private Direction determineEastOrWest(boolean isNegative)
    {
        return isNegative ? Direction.West : Direction.East;
    }

    private Direction determineNorthOrSouth(boolean isNegative)
    {
        return isNegative ? Direction.South : Direction.North;
    }
}
